wincd.man



WINCD(1)						 WINCD(1)


NAME
       wincd  -	 a  daemon  which gets images from the WinCam.One
       camera.

SYNOPSIS
       wincd [-options]

DESCRIPTION
       wincd periodically gets and saves images from the  digital
       camera known as the WinCam.  It is a 640x492 capable color
       digital camera, interfaced to  a	 computer  via	a  serial
       cable.	See  http://www.wincam.com  for	 more information
       about this camera.

       Images grabbed by wincd are dumped in PGM or PNM format to
       a  specified  file.   An	 optional script may be specified
       which will run after  the  image	 is  saved.   Optionally,
       images may be saved only if they are determined to be dif-
       ferent than previous images.  "Difference" in this case is
       determined  by  comparing  reduced  versions  of	 the  two
       images: if a 64x48 version of the image has fewer  than	M
       pixes  which  differ  by more than R grey levels, then the
       images are considered to be the "same".	(These limits are
       settable with -r and -m.)

       wincd  controls	image  exposure automatically, though the
       WinCamBrightness configuration variable	can  be	 used  to
       control	the  brightness	 goal.	(See the winc manual page
       for information on configuration variables.)


LOCKING
       Under OS/2, wincd.exe locks the com port while in use.

OPTIONS
       -h     prints the usage message

       -p pollseconds
	      The  loop interval is the amount of time wincd will
	      pause during iterations of the  image  poll/compare
	      loop.   Default  is 10.  This is the interval wincd
	      will pause after it does not take a picture.

       -w waitseconds
	      The wait interval is the amount of time wincd  will
	      pause  when it does take a picture.  Default is 60.

       -n nsave
	      wincd will keep "small" versions of images that  it
	      has  saved  in  memory,  and  compare freshly taken
	      images to these.	If it determines that a new image
	      is  a duplicate of one that has already been saved,
	      it will not save the new one.  The -n option deter-
	      mines the length of this history, which defaults to
	      6.  If simple "take  a  picture  every  N	 seconds"
	      behavior is desired, set nsave to 0 and use "-w N".
	      If a comparison history is desired,  set	nsave  to
	      the  length  of  the history.  Use -p to choose how
	      often wincd will look for changed images, and -w to
	      "rate-limit"  the frequency of real pictures taken.

       -r threshold

       -m maxdiff
	      Successive images are compared (in  reduced,  64x48
	      form)   until  an	 "interesting"	image  is  found.
	      "Interesting" is defined	as  "different	than  the
	      last",  and  that	 is  defined  as having more than
	      MAXDIFF percent of the pixels differ  by	at  least
	      THRESHOLD	 percent of the total grey levels.  MAXD-
	      IFF is expressed as a percentage of image area, and
	      THRESHOLD	 is  expressed as a percentage of maximum
	      pixel value.  When the draft-quality  image  passes
	      this  criterion,	then  a full-size image is taken.
	      (So actually, the premise that comparisons are done
	      on reduced versions of images is false -- actually,
	      the reduced and fullsize images come from two sepa-
	      rate exposures.)

       -f filename
	      Images  grabbed by wincd are saved in the specified
	      filename.	 It is	recommended  that  this	 file  be
	      given  the  suffix  ".ppm" or ".pgm", since this is
	      the conventional suffix for images in  PNM  or  PGM
	      format.	(PPM for color, PGM for greyscale) (These
	      formats are specified as part of the widely  avail-
	      able  PBMPLUS  or "netpbm" packages.)  

             (OS/2 Note:  The Independent JPEG Group's JPEG Software 
              v 6a is currently at:
                ftp://hobbes.nmsu.edu/old/os2/graphics/jpeg6a_os2a.zip

             The GBM Utils also deal with the PNM format - they can
             currently be found at:
	        ftp://hobbes.nmsu.edu/old/os2/graphics/gbm.zip

             As the Hobbes archive is re-organizing as this is being 
             written, expect this to change. )



              The  filename  defaults  to    "wincd.pnm".     The  
              modification  time  of the file  will be set to the 
	      time at which the picture  was snapped, rather than 
              being  set  by  the  system  to  the time  at which 
              the last part  of the  file was written.  (The time 
              difference can  be tens of seconds.)

       -s scriptname
	      After  an	 image is saved, the specified script (if
	      any) will be run.	 It is handed the filename of the
	      image  as	 its  only  argument.  

       -c     Grabs color images.  This is the default action.

       -g     Grabs  greyscale	images,	 instead of color images.
	      Greyscale images are 8 bits deep,	 with  a  maximum
	      grey  level of 255.  Color images are 24 bits deep,
	      with 255 values for each of red, green,  and  blue.
	      The  default  size  of  either  type  of	image  is
	      640x492.

       -l lighting
	      Affect the color balance of  the	image.	 See  the
	      COLOR  BALANCE  section of the winc manual page for
	      more information.

       -2     Normally images are captured in a single	scan:  in
	      this  mode,  of  the  492 lines in a default image,
	      only the even rows are "real":  the  odd	rows  are
	      created  by duplicating the row above.  If the sub-
	      ject matter is still enough, the -2 option  can  be
	      used to force a second scan about 1/8th of a second
	      later.  This will fill in the odd	 rows  with  real
	      data, and yield a higher quality picture.

       -z N   Sets  the	 image reduction factor to 1, 2, 4.  This
	      might be done when the full imagesize or resolution
	      is not wanted or needed.	The values 1, 2, 4, and 8
	      represent linear scaling of both the horizontal and
	      vertical	axes.  Thus "-z 4" gives an image that is
	      160x123 rather than the default 640x492.	(So actu-
	      ally,  the arguments should really be 1, 4, 16, and
	      64.)

       -e NN or -E NN
	      Sets exposure in milliseconds(-e)	 or  microseconds
	      (-E).   The  shortest  possible  exposure	 time  is
	      1/100,000th of a second, the longest  is	a  little
	      over  1.3 seconds.  If neither option is used, winc
	      will try several exposures until an  average  pixel
	      brightness  goal	is  reached, which is set via the
	      WinCamBrightness variable.

       -a     Sound an audible alarm when the picture is actually
	      snapped.	The process of "finding" the camera, tak-
	      ing the  picture,	 and  downloading  the	image  is
	      fairly lengthy.  Hearing a small beep when the pic-
	      ture has been taken can be helpful.

       -i IDSTRING
	      Specify the "name" of the camera to be used.   This
	      name  will  affect  the  way in which configuration
	      information will be looked up.  See the section  on
	      MULIPLE  CAMERAS	on  the winc manual page for more
	      information.

       -t tracestring
	      Turns on program tracing.	 See the winc manual page
	      for more information.

STARTUP
       winc will load configuration  information  from	the  file
       $ETC/winc.rc. $ETC, for  those of  you who don't  know, is
       the  directory pointed  to by  the "SET ETC=" line in your
       config.sys file. This  pathname can be overridden with the
       $WINC_RC environment  variable.	See  the manual  page for 
       the "winc" command for more details.

COPYING and COPYRIGHT
       This software is Copyright (C) 1996, by Paul G. Fox.  Spe-
       cific  image  processing	 code  is  Copyright (C) 1996, by
       StarDot Technologies.

       This program is free software;  you  can	 redistribute  it
       and/or modify it under the terms of the GNU General Public
       License as published  by	 the  Free  Software  Foundation;
       either  version	2 of the License, or (at your option) any
       later version.

       This program is distributed in the hope that  it	 will  be
       useful, but WITHOUT ANY WARRANTY; without even the implied
       warranty of MERCHANTABILITY or FITNESS  FOR  A  PARTICULAR
       PURPOSE.	  See  the  GNU	 General  Public License for more
       details.

       You should have received a copy of the GNU General  Public
       License along with this program; if not, write to the Free
       Software Foundation, Inc., 675  Mass  Ave,  Cambridge,  MA
       02139, USA.

DEBTS
       wincd was created with the help of proprietary information
       belonging to StarDot Technologies, makers of  the  WinCam.
       Contact them at help@wincam.com for information on obtain-
       ing a copy of that information.

AUTHORS
       wincd and the winc library were created by  Paul	 G.  Fox,
       pgf@foxharp.boston.ma.us.

       Port to OS/2 by Derek J Decker, derek@decker.net

BUGS
       Oh well.















								4




ChangeLog

Changes for winc-36:
    961016
	+ README: mention the mailing list.  subscribe at:
	    pgf-winc-request@foxharp.boston.ma.us
    960929
	+ winc.c wincd.c xwinc.c: added missing break in getopt switch.

Changes for winc-35:
    960926
	+ find.c: zero the cam struct after mallocing.  current symptom
	  was an uninitialized "starfield" element that was causing
	  core dumps.

	+ wincd.c wincd.1: add -e and -E for controlling exposure.

Changes for winc-34:
    960910
	+ wincd.c winc.1: apply gamma function, update man page.

	+ vignette.c winc.c wincam.h:  implemented winc_gamma(), which
	  applies a gamma function to the image.  the default gamma value
	  is 100 (representing 1.0 in the actual equation).

    960828
	+ xwinc.1 winc.1: clarifications, syntax errors.

	+ view.c: if avg becomes zero while auto-exposing, set to high
	  default, since it might be zero due to not enough light.  this
	  is probably only the case when the lens cap is on, but possibly
	  at night?

	+ winc.c: choose a goal when doing starfield shots.  let's call it
	  50.  this keeps the winc.rc file from influencing the results.

	+ winc.1: a few more comments about creating the starfield image.
	  the home camera wanted -S 20, the office camera wanted -S 45.
	  this is either a difference in cameras, or a difference in
	  temperature -- the office camera is on top of a warm monitor,
	  the one at home beside a cool wall.
    
Changes for winc-33:
    960828
	+ pnm.c image.c convert.c find.c private.h winc.c winc.1:  changes
	  to support starfield reference images.

	+ Makefile starfield.c: new file containing code for reading and using
	  the starfield reference image.

Changes for winc-32:
    960826
	+ wincd.c winc.c wincd.1 winc.1:  reset the mod time of the written
	  image to be the time at which the image was snapped.

	+ winc.c: added "-o filename" to replace stdout as image destination.

    960823
	+ winc.c: added -S, for saving starfield pbm file.  nothing uses
	  this yet.

	+ pnm.c pnm.h: added winc_pbm_output(), for mono bitmap output

Changes for winc-31: (mostly camera id support -- for naming cameras)
    960822
	+ winc.c: usage message is now split in two -- it was way
	  too long.

	+ winc.c: added -s speed option for setting baud rate.

	+ winc.1: rearranged options, added section on baud rate and
	  timeouts.

	+ grabber.sh: add disclaimer about how wincd/wincd_script do
	  everything (?) this does, perhaps better.

    960821
	+ convert.c winc.1:  added WinCamColor variable for specifying
	  default color balance.

	+ winc.1 wincd.1 xwinc.1: documentation of -i ID option.

    	+ winc.c: unlock camera before normal closing, to suppress
	  superfluous (and time-consuming) flush.

	+ config.c: add support $WinCamId as a prefix for other config
	  variables.  i.e. if WinCamId is set to "kitchen", then when
	  we are asked for to look up "WinCamDevice", we will check for
	  "kitchenWinCamDevice" first.

	+ mergeenv.c: added "safeputenv()", which mallocs before doing
	  a putenv() on its args.

	+ winc.c wincd.c xwinc.c: add support for "-i idstring", for
	  setting the WinCamId variable.  this lets you use multiple
	  cameras on the same system, all configured from the same
	  config file.

    	+ find.c: include string.h for strerror() decl.

    	+ convert.c: log the color balance choices made.

Changes for winc-30:  (mostly modem support, also wincd bugfix)
    960821
	+ find.c: if WinCamModemChat config variable is non-null, then
	  the line is tested for a modem.  if one is found, the contents
	  of that config variable are run via "sh -c", with the tty port
	  as stdin and stdout.  the command is intended to be a chat
	  program, which will negotiate with and dial the modem.

	+ protocol.c, private.h: added winc_check_for_modem() routine,
	  to distinguish modem ports from camera ports.

	+ comms.c: open the tty O_NDELAY, to ensure it opens.

	+ wincd.c:  change DEFAULT_MAXDIFF to just 1% (from 4%).  i
	  mis-translated the old numbers when i converted the code to use
	  percentages.  it _used_ to use just 10 out of 3072 (64x48)
	  pixels, which is .0032, or .32%, not 3.2%.

    960818
	+ wincd_script: change from /bin/sh to /bin/bash.

	+ winc.1 wincd.1: fix syntax error.

Changes for winc-29:	(mostly implementation of color-balance controls)

    960818
    	+ winc.c winc.1: no more -m option.  -e or -E forces manual
	  exposure, -b allows setting goal, which is passed as negative
	  exposure setting to the snap routines.

	+ image.c convert.c wincam.h view.c image.c:  exposure is passed as
	  int not unsigned, so goal can come in as negative exposure value.

	+ view.c: change default brightness to 50%.

	+ wincd_script:  make it a little smarter about its args.

	+ xwinc.c: brightness default now comes from config file.

	+ rationalize different default brightness settings, add -b flag
	  to winc.c for controlling brightness goal.

    	+ color.c: the problem that green-level is "off by 59%" is
	  now taken care of in here.  blue, red, and green fix values
	  are now all based on 128 externally.

	+ convert.c winc.c wincd.c wincam.c:  API change: added "image
	  adjustment" struct to arg list, added "winc_image_adjustments()"
	  routine to fill it in, based on user-provided color-balance string,
	  and widen and vignette choices.

	+ winc.c wincd.c winc.1 wincd.1:  added -l option, which lets
	  one set lighting (color-balance) specifications.

	+ winc.c: re-enabled the widen (-n) and vignette (-r) options.

    960817
    	+ convert.c: commentary about color balancing values.

	+ wincd.c wincd.1: change -l SECS to -p SECS (for "poll") to
	  free up -l for "lighting" control.

Changes for winc-28:	(mostly cosmetic renaming/relayering)
    960816
	+ comms.c config.c convert.c diags.c find.c locks.c private.h
	  protocol.c view.c vignette.c winc.1 wincam.h wincd.c xwinc.c:
	  mostly cosmetic.  renamed some internal routines to show their

src/wincd.c

/*
 *
 * winc -- a minimal "daemon" that collects and saves interesting images
 *
 * Copyright (C) 1996, Paul G. Fox
 * This program is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by the
 * Free Software Foundation; either version 2 of the License, or (at your
 * option) any later version.
 * 
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 * This software was created with the help of proprietary information
 * belonging to StarDot Technologies.
 *
 * $Header: E:/winc/RCS/wincd.c 1.1 1997/03/01 03:44:14 Derek Exp Derek $
 */

#include <unistd.h>
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <signal.h>
#include <sys/types.h>
#include <sys/wait.h>
#include <time.h>
#include <utime.h>
#include <sys/stat.h>
#include "wincam.h"
#include "pnm.h"
#include "trace.h"

#define DEFAULT_POLLSECS    10
#define DEFAULT_WAITSECS    60
#define DEFAULT_THRESHOLD   25	/* percent of max pixel value */
#define DEFAULT_MAXDIFF     1	/* percent of total pixels */
#define DEFAULT_NSAVE	    6
#define DEFAULT_SCANTYPE    1
#define DEFAULT_FRACTION    1
#define DEFAULT_FILENAME    "wincd.pnm"
#define NSAVEMAX 20

/*
 * from mergeenv.c -- there's no convenient header file in which
 * this appears.
 */
int	safeputenv(char *name, char *value);

void snapper_func(void *timep);
void audible_snapper_func(void *timep);
void reset_mod_time(char *filename, time_t mtime);
void quitcatcher(int sig);

/*
 * global camera handle, so it can be closed/flushed from a signal handler
 */
cam_t Cam = 0;		/* camera "handle" */

/*
 * for getopt()
 */
extern char *optarg;
extern int optind, opterr, optopt;


void
usage(char *progname)
{
    fprintf(stderr, 
    "usage: %s -{flags}  where:\n"
    "   -h gives usage message\n"
    "   -f FILE sets filename into which to store image [wincd.pnm]\n"
    "   -s SCRIPT - runs when file saved, FILE is only arg. [no default]\n"
    "   -p SECS - sleep interval when polling to see if image changed [10]\n"
    "   -w SECS - sleep interval to wait after image has been snapped [60]\n"
    "   -n N - how many images for which to track history [6]\n"
    "	(use \"-n 0\" for simple save-every-time operation)\n"
    "    -r N - change in pixel (as %% of max) to be called changed [25]\n"
    "    -m N - percent of pixels changed for image to be different [4]\n"
    "   -z N - image reduction factor (one of 1, 2, 4, or 8) [1]\n"
    "   -2 to double-scan an image\n"
    "   -g grabs a greyscale image\n"
    "   -c grabs a color image (default)\n"
    "   -l color-adjustment\n"
    "   -a causes an audible alert (beep) when image is exposed\n"
    "   -i IDSTRING assigns a prefix for all config info looked up\n"
    "   -t STR sets tracing: see winc usage message\n"
    "", progname);
}

/*
 * release the camera while we sleep
 */
void
camsleep(cam_t cam, int secs)
{
    winc_unlock(cam);
    sleep(secs);
    winc_lock(cam);
}

int
main(int argc, char *argv[])
{
    char *progname;
    int c;

    int grey = 0;
    int pollsecs = DEFAULT_POLLSECS;	/* seconds to pause while looping */
    int waitsecs = DEFAULT_WAITSECS;	/* seconds to pause after snapping */
    int scantype = DEFAULT_SCANTYPE;	/* single- or double- scanned */
    int fraction = DEFAULT_FRACTION;	/* full image? what fraction? */
    char *filename = DEFAULT_FILENAME;	/* where to put the image */
    char *scriptname = 0;		/* script to run on saved image */
    char *color_adj_string = 0;
    int audible_alert = 0;

    int zoom = 1;		/* no viewfinder zoom by default */
    int nsave = DEFAULT_NSAVE;
    int thresh = DEFAULT_THRESHOLD;
    int maxdiff = DEFAULT_MAXDIFF;

    int exp = 0;
    int avg = 0;
    int blackadj;
    int same, sameas = 0;
    unsigned char *saveimage[NSAVEMAX];
    unsigned char *t;
    FILE *outf;
    time_t snaptime;		/* for resetting image grab time */
    int rows, cols;
    int pid = -1;
    int i;
    struct winc_image_adjust image_adjust;

    unsigned char image[640*492];	/* the biggest imagemap we'll need */
    unsigned char convimage[3*640*492]; /* the biggest converted imagemap */

    /* 
     * what program is this?
     */
    progname = strrchr(argv[0], '/');
    if (progname)
	progname++;
    else
	progname = argv[0];

    _fsetmode(stdout,"b");


    /* 
     * set up the trace and error facilities
     * register the output callbacks, and initialize the tracing class
     */
    register_error_func((vprintffunc)vfprintf, (void *)stderr);
    register_trace_func((vprintffunc)vfprintf, (void *)stderr);
    set_trace_class("");

    /* 
     * process arguments
     */
    while ((c = getopt(argc, argv, "2hgcae:n:w:l:p:s:z:t:f:r:m:i:")) != EOF) {
	switch (c) {
	case 'h':
	    usage(progname);
	    exit(0);
	case 'z':
	    zoom = atoi(optarg);
	    fraction = zoom;
	    break;
	case 'e':   /* exposure (or initial exposure guess), in milliseconds */
	case 'E':   /*	or microseconds */
	    exp = atoi(optarg);
	    if (c == 'e')
		exp *= 1000;	    /* convert to microseconds */
	    break;
	case 't':
	    set_trace_class(optarg);
	    break;
	case 'a':
	    audible_alert  = 1;
	    break;
	case 'f':
	    filename = optarg;
	    break;
	case 's':
	    scriptname = optarg;
	    break;
	case 'n':
	    nsave = atoi(optarg);
	    break;
	case 'g':
	    grey = 1;
	    break;
	case 'c':
	    grey = 0;	/* this is the default in any case */
	    break;
	case 'l':
	    color_adj_string = optarg;
	    break;
	case '2':
	    scantype = 2;
	    break;
	case 'p':
	    pollsecs = atoi(optarg);
	    break;
	case 'w':
	    waitsecs = atoi(optarg);
	    break;
	case 'r':
	    thresh = atoi(optarg);
	    break;
	case 'm':
	    maxdiff = atoi(optarg);
	    break;
	case 'i':
	    if (safeputenv("WinCamId", optarg) == -1) {
		fprintf(stderr, "%s: couldn't set WinCamId\n", progname);
	    }
	    break;
	default:
	    fprintf(stderr, "%s: unknown option '%c'\n", progname, c);
	    usage(progname);
	    exit(1);
	}
    }

    if (optind != argc) {
	fprintf(stderr, "%s: extra arguments\n", progname);
	usage(progname);
	exit(1);
    }


    if (!filename || !filename[0]) {
	fprintf(stderr, "%s: must give -f filename\n", progname);
	exit(1);
    }

    if (nsave > NSAVEMAX) {
	errormsg(__FILE__ ": nsave too big -- fix NSAVE and recompile\n");
	exit(1);
    }

    if (audible_alert != 0)
	winc_register_alerter(audible_snapper_func, (void *)&snaptime);
    else
	winc_register_alerter(snapper_func, (void *)&snaptime);

    winc_image_adjustments(&image_adjust, color_adj_string, 1, 1);

    for (i = 0; i < nsave; i++) {
	saveimage[i] = (unsigned char *)malloc(48 * 64);
	if (!saveimage[i]) {
	    errormsg(__FILE__ ": can't malloc saveimages\n");
	}
	memset(saveimage[i], 0, 48*64);
    }

refind:
    Cam = winc_find(1);

    if (!Cam) {
	errormsg(__FILE__ ": couldn't find camera\n");
	exit(1);
    }

    (void)signal(SIGHUP, quitcatcher);
    (void)signal(SIGINT, quitcatcher);
    (void)signal(SIGTERM, quitcatcher);

    if (winc_black_offset(Cam, &blackadj, 0) != OK) {
	errormsg(__FILE__ ": failed to get black adjustment\n");
	goto refind;
    }

    while (1) {

	winc_lock(Cam);

	/* take a viewf image, looping to get exposure */
	if (winc_viewfinder_snap(Cam, 1, -1, -1, image,
			(exp == 0) ? 8 : 1, &exp, &avg, 0, 0) != OK) {
	    errormsg(__FILE__ ": failed to snap viewfinder image\n");
	    winc_close(Cam);	/* unlocks */
	    sleep(pollsecs);
	    goto refind;
	}

	winc_unlock(Cam);

	TRACE("xe", __FILE__ ": exposure %d.%03d sec (%d usec)\n",
	    exp/1000000, (exp%1000000)/1000, exp);

	if (nsave > 0) {
	    /* compare to last few images */
	    same = 0;
	    for (i = 0; i < nsave; i++) {
		int s = winc_compare_image(image, saveimage[i], 48, 64,
			    maxdiff, thresh, 15, 0);
		if (s && !same) sameas = i;  /* overwrite the first match */
		TRACE("x", __FILE__ ": image %d compare %ssame\n", i, 
					s ? "" : "not ");
		same = same || s;
	    }


	    if (same) {
		memcpy(saveimage[sameas], image, 48 * 64);
		sleep(pollsecs);
		/* try to pick up the last script that ran, just to
		    catch zombies earlier.  we don't block, though */
		if (pid != -1) {
		    if (waitpid(pid, 0, WNOHANG) == pid)
			pid = -1;   /* got it -- quit trying */
		}
		continue;
	    }

	    /* save new image, rotate the previous */
	    t = saveimage[nsave-1];
	    for (i = nsave-1; i > 0; i--)
		saveimage[i] = saveimage[i-1];
	    saveimage[0] = t;
	    memcpy(saveimage[0], image, 48 * 64);
	}

	winc_lock(Cam);

	TRACE("x", __FILE__ ": snapping full image\n");
	winc_snap_and_process(Cam, exp, image, convimage, &rows, &cols,
				!grey, scantype, fraction, &image_adjust);

	winc_unlock(Cam);

	/* pick up the last script that ran, before overwriting the
	   file it was processing */
	if (pid != -1) {
	    TRACE("x", __FILE__ ": reaping script process %d\n", pid);
	    if (waitpid(pid, 0, 0) != pid) /* it had to be, right? */
		TRACE("x", __FILE__ ": got wrong process? %d\n", pid);
	    pid = -1;
	}

	/* process it, widen it, write it out, to a known name. spawn
		a script to deal with it, passing filname as arg. */

	outf = fopen(filename, "wb");
	if (!outf) {
	    errormsg(__FILE__ ": can't open file \"%s\"\n", filename);
	    exit(0);
	}

	winc_gamma(0, convimage, rows, cols, !grey);

	if (grey)
	    (void)winc_pgm_output(convimage, rows, cols, 255, outf);
	else
	    (void)winc_pnm_output(convimage, rows, cols, 255, outf);

	fclose(outf);
	reset_mod_time(filename, snaptime);

	if (scriptname) {
	    TRACE("x", __FILE__ ": spawning script %s on %s\n",
	    					scriptname, filename);
	    if (system(scriptname) == -1) { /* child */
		winc_close(Cam);
		execlp(scriptname, scriptname, filename, 0);
		errormsg(__FILE__ ": exec of %s failed\n", scriptname);
		exit(1);
	    } 
	}

	/* wait for specified time after snapping a picture */
	if (waitsecs != 0)
	    sleep(waitsecs);

    }
    if (Cam)
	winc_close(Cam);
    return 0;
}

/*
 * the snapper funcs are "called back" by the winc library when the
 * picture is actually snapped.  this lets us a) make a noise, and b)
 * record the time
 */
void
snapper_func(void *timep)
{
    *(time_t *)timep = time(0);
}

void
audible_snapper_func(void *timep)
{
    putc('\a', stderr);
    *(time_t *)timep = time(0);
}

/*
 * attempt to reset the modification time of the given file to mtime.
 * this is used to set the time on the output file to the time at which
 * the image was snapped, rather than when it was written, since processing
 * can take some time.
 */
void 
reset_mod_time(char *filename, time_t mtime)
{
    struct stat stb;
    struct utimbuf utb;
    if (stat(filename, &stb) == -1)
    	return;

    utb.actime = stb.st_atime;
    utb.modtime = mtime;
    (void)utime(filename, &utb);
}

void
quitcatcher(int sig)
{
    if (Cam)
    	winc_close(Cam);
    exit(1);
}

